demo/src/main/scala/demo/DemoMain.scala

```diff
@@ -13,6 +14,158 @@ import scala.scalajs.js.typedarray.{ArrayBuffer, Uint8Array}
+    ).render)
+
+    var notes = Vector.empty[String]
+    var progress = ""
+
+    def updateLogBox() = {
+      loggerBox.textContent = (
+        Seq(
+          "--------Remaining Events--------",
+          progress,
+          "---------Current Notes----------"
+        ) ++
+        notes.take(1) ++
+        Seq(
+          "---------Earlier Notes----------"
+        ) ++
+        notes.slice(1, 10)
+      ).mkString("\n")
+    }
+
+    uploadFile.onchange = (e: dom.Event) => {
+      val reader = new dom.FileReader()
+      reader.readAsArrayBuffer(uploadFile.files.item(0))
+      reader.onload = (e: UIEvent) => {
+        val array = new Uint8Array(reader.result.asInstanceOf[ArrayBuffer])
+        val contents = (for (i <- 0 until array.length) yield array.get(i).toByte).toArray
+        dom.console.log(contents.length)
+        fastparse.MidiParse.midiParser.parse(contents) match{
+          case Parsed.Success(midi, index) =>
+            metadataBox.textContent =
+              s"""Successfully parsed ${index} bytes
+                  |Midi Format: ${midi.format}
+                  |Tick Division: ${midi.tickDiv}
+                  |${midi.tracks.length} Tracks:
+                  |${midi.tracks.zipWithIndex.map{case (t, i) => s"Track $i: ${t.length} events"}.mkString("\n")}
+                 """.stripMargin
+            playMusic(midi, {x => notes = x +: notes; updateLogBox()}, {x => progress = x; updateLogBox()})
+          case Parsed.Failure(lastParser, index, extra) =>
+            metadataBox.textContent = "Failed to parse midi\n" + extra.traced.trace
+        }
+      }
+    }
+
+  }
+  val notes = Vector("C","C#","D","D#","E","F","F#","G","G#","A","A#","B")
+  lazy val allNotes = for(octave <- 0 to 10; note <- notes) yield (note, octave)
+  def makeNote(note: String, octave: Int) = {
+    js.Dynamic.newInstance(js.Dynamic.global.Audio)(
+      js.Dynamic.global.Synth.generate(0, note, octave - 2, 10)
+    )
+  }
+  def playMusic(midi: fastparse.Midi,
+                noteLogger: String => Unit,
+                progressLogger: String => Unit) = {
+    import fastparse.Midi._, MidiData._
+    case class TrackInfo(var savedTicks: Int, var track: List[(Int, Midi.TrackEvent)]){
+      def tillNext = track.head._1 - savedTicks
+    }
+    noteLogger("Initializing Musical Notes...")
+    val notes =
+      midi.tracks
+          .flatten
+          .collect{case (_, MidiEvent(channel, NoteOn(note, velocity))) => (channel, note)}
+          .distinct
+          .map{ case (channel, note) => ((channel, note), makeNote(allNotes(note)._1, allNotes(note)._2)) }
+          .toMap
+
+    val allTracks = midi.tracks.map(_.toList).map(TrackInfo(0, _))
+    // 120bpm, in microseconds-per-quarter-beat
+    var currentTempo = 1000000 * 120 / 60 / 4
+
+    def tick(): Unit = {
+
+      val logs = collection.mutable.Buffer.empty[String]
+      progressLogger(allTracks.map(_.track.length).mkString(", "))
+      while ({
```

Comment:
Well, it's not strictly necessary, but it's convenient. `do-while` doesn't work because the things you define inside the `do` block aren't available in the conditional when deciding whether or not to continue. In a more traditional language I would use `while(true)` and `break`, but `break` in Scala is funky (it throws exceptions!), and nobody really uses it, so I'd rather avoid it. I could re-write it as tail-recursive, but that makes the code a lot messier than a simple loop